### ISSUE TITLE
Properly document the internal conversion of string labels to ints/floats and possible edge cases

### ISSUE THREAD

**Issue by mulhod** [EducationalTestingService/skll#436]
This is a bit of a strange issue, but I have worked with a data-set that has labels that look like `2`, `2.1`, `2.21`, `2.3`, and `2.31`. Using `LinearSVC` I was expecting the predicted labels to look like those labels. However, due to the use of `safe_float` in the various reader classes, `"2"` gets converted to an `int` `2`, `"2.21"` a `float` `2.21`, etc. Then, after a Python list is constructed of these `int`s and `float`s, this list is converted to a `numpy` array [here](https://github.com/EducationalTestingService/skll/blob/3f65d27cff13cfcba7bbf2b30d893cd074642913/skll/data/readers.py#L234). So, now we have a `numpy` array of `float`s: `np.array([2.0, 2.2, 2.21, ...])`. The resulting classification model predicts these types of labels, which is unexpected and can cause lots of problems down the road. It seems to me that, if a classification algorithm is being used, `safe_float` shouldn't even be applied: just leave the labels alone! This would mean that data-sets would depend on what type of algorithm is being used, classification or regression. And that, in turn, would cause issues for the way we allow a list of classifiers/regressors to be specified in the experiment configuration files: We would either have to not allow a combination of such algorithms or read in the data twice, once for each type of algorithm.

**Comment by desilinguist** [comment]
I am not really sure I understand the issue. The same thing would happen if you have labels that were just "2", "3", and "4". You would get an array of floats but it's a set of discrete labels so classification still works as expected, no?

**Comment by desilinguist** [comment]
Why not use the [`class_map`](https://skll.readthedocs.io/en/latest/run_experiment.html#class-map-optional) functionality that's already in SKLL. This should solve your problem perfectly.

**Comment by mulhod** [comment]
I don't disagree. It is unfortunate data. And, yes, I think I can use `class_map` or something else. I've definitely figured out a way around this, but it just seems to me that the behavior is unexpected. Maybe there should be a warning if `safe_float` results in a list of labels that are of different types?

**Comment by desilinguist** [comment]
I think that's the whole point of `safe_float` - to convert strings to either ints or floats because those are more efficient data types and it doesn't really make a difference to the underlying ML algorithm either way.

**Comment by mulhod** [comment]
I should say that this would only be for classification where one might want to know it if the labels are being transformed in some way. It would be easy to think that they are string values and won't change.

**Comment by desilinguist** [comment]
My argument is that I don't think it's reasonable to expect labels like "2.21" to be treated as strings.

**Comment by mulhod** [comment]
They are, though, if there are other strings.

```
In [8]: np.array([safe_float(x) for x in ["2", "3", "4"]])
Out[8]: array([2, 3, 4])

In [9]: np.array([safe_float(x) for x in ["2", "2.2", "2.21"]])
Out[9]: array([2.  , 2.2 , 2.21])

In [10]: np.array([safe_float(x) for x in ["x", "2", "2.2", "2.21"]])
Out[10]: array(['x', '2', '2.2', '2.21'], dtype='<U4')
```

**Comment by mulhod** [comment]
Anyway, I get your point. My point is that somebody did think it was reasonable, unfortunately, and it caused some confusion and didn't get noticed until well after it happened, even with lots of eyes on it.

**Comment by desilinguist** [comment]
How about this? I am not sure a warning in the code is warranted here but perhaps something in the documentation? Would you mind thinking about where that should go?

**Comment by desilinguist** [comment]
Addressed by #476.
